lib/tools/apk-utils.js

```diff
@@ -474,12 +474,81 @@ apkUtilsMethods.getDeviceLocale = async function () {
 
 /**
  * Set the locale name of the device under test.
+ * Compatibility for appium-android-driver 1.32.4-
  *
  * @param {string} locale - The name of the new device country.
  */
 apkUtilsMethods.setDeviceLocale = async function (locale) {
-  // this method is only used in API >= 23
-  await this.setDeviceSysLocale(locale);
+  let split_locale = locale.split("-");
+  await this.setDeviceLanguageCountry(split_locale[0], split_locale[1]);
+};
+
+/**
+ * Set the locale name of the device under test.
+ *
+ * @param {string} language - Language. e.g. en, ja
+ * @param {string} country - Country. e.g. US, JP
+ */
+apkUtilsMethods.setDeviceLanguageCountry = async function (language, country) {
+  let haveLanguage = language && typeof language === "string";
+  let haveCountry = country && typeof country === "string";
+  if (!haveLanguage && !haveCountry) {
+    return;
+  }
+  let changed = false;
+  let apiLevel = await this.getApiLevel();
+
+  if (apiLevel < 23) {
+    let curLanguage = await this.getDeviceLanguage();
+    let curCountry = await this.getDeviceCountry();
+    if (haveLanguage && language !== curLanguage) {
+      await this.setDeviceLanguage(language);
+      changed = true;
+    }
+    if (haveCountry && country !== curCountry) {
+      await this.setDeviceCountry(country);
+      changed = true;
+    }
+  } else if (apiLevel < 24) { // API == 23
+    let curLocale = await this.getDeviceLocale();
+    let locale;
+    if (!haveCountry) {
+      locale = language.toLowerCase();
+    } else if (!haveLanguage) {
+      locale = country;
+    } else {
+      locale = `${language.toLowerCase()}-${country.toUpperCase()}`;
+    }
+
+    log.debug(`Current locale: '${curLocale}'; requested locale: '${locale}'`);
+    if (locale !== curLocale) {
+      await this.setDeviceSysLocale(locale);
+      changed = true;
+    }
+  } else { // API >= 24
+    let curLocale = await this.getDeviceLocale();
+    let localeLang;
+    let localeCountry;
+    if (!haveCountry) {
```

Comment:
Can we refactor the code to avoid duplicating this condition for api == 23/api > 23 cases?

Reply:
BTW, we can **remove** the `api == 23` case and use `setDeviceSysLocaleViaSettingApp` instead since I've used `setDeviceSysLocaleViaSettingApp` logic to change device locale for all devices in my company and the devices include api 23.

^ is just my thinking and what do you think @mykola-mokhnach ? (It's okay if you have no exact opinion, though)
But I also think it's better to keep current behaviour, using adb command, for the `api == 23/api > 23` since they work fine so far. 🤔

Reply:
The logic is ok there, no need to remove anything. I just wanted to optimise some redundant code and the recent state of this code block makes me happy :)